kernel/src/hil/hasher.rs

```diff
@@ -0,0 +1,83 @@
+use crate::utilities::leasable_buffer::LeasableBuffer;
+use crate::ErrorCode;
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+///
+/// 'L' is the length of the 'u8' array to store the hash output.
+pub trait Client<'a, const L: usize> {
+    /// This callback is called when the data has been added to the hash
+    /// engine.
+    /// On error or success `data` will contain a reference to the original
+    /// data supplied to `add_data()`.
+    fn add_data_done(&'a self, result: Result<(), ErrorCode>, data: &'static mut [u8]);
+
+    /// This callback is called when a hash is computed.
+    /// On error or success `hash` will contain a reference to the original
+    /// data supplied to `run()`.
+    fn hash_done(&'a self, result: Result<(), ErrorCode>, hash: &'static mut [u8; L]);
+}
+
+/// Computes a non-cryptographic hash over data
+///
+/// 'L' is the length of the 'u8' array to store the hash output.
+pub trait Hasher<'a, const L: usize> {
+    /// Set the client instance which will receive `hash_done()` and
+    /// `add_data_done()` callbacks.
+    /// This callback is called when the data has been added to the hash
+    /// engine.
+    /// The callback should follow the `Client` `add_data_done` callback.
+    fn set_client(&'a self, client: &'a dyn Client<'a, L>);
+
+    /// Add data to the hash block. This is the data that will be used
+    /// for the hash function.
+    /// Returns the number of bytes parsed on success
+    /// There is no guarantee the data has been written until the `add_data_done()`
+    /// callback is fired.
+    /// On error the return value will contain a return code and the original data
+    fn add_data(
+        &self,
+        data: LeasableBuffer<'static, u8>,
+    ) -> Result<usize, (ErrorCode, &'static mut [u8])>;
+
+    /// Request the hardware block to generate a Hasher and stores the returned
+    /// hash in the memory location specified.
+    /// This doesn't return any data, instead the client needs to have
+    /// set a `hash_done` handler to determine when this is complete.
+    /// On error the return value will contain a return code and the original data
+    /// If there is data from the `add_data()` command asyncrously waiting to
+    /// be written it will be written before the operation starts.
+    ///
+    /// If an appropriate `set_mode*()` wasn't called before this function the
+    /// implementation should try to use a default option. In the case where
+    /// there is only one hash supported this should be used. If there is no
+    /// suitable or obvious default option, the implementation can return an
+    /// error with error code ENOSUPPORT.
+    fn run(&'a self, hash: &'static mut [u8; L]) -> Result<(), (ErrorCode, &'static mut [u8; L])>;
+
+    /// Clear the internal state of the engine.
+    /// This won't clear the buffers provided to this API, that is up to the
+    /// user to clear.
+    fn clear_data(&self);
+}
+
+pub trait Sha224 {
```

Comment:
Why are cryptographic hashes in a HIL for non-cryptographic hashes? It seems like there should instead be a SipHash trait. There are different parameters for SipHash: should those be settable?

Reply:
And also: doesn't the `set_mode` style of selecting a hash function collide with how the output length is encoded as a const generic? I.e. `Hasher<64>` is a valid type specification for some SHA2-512 hash algorithm implementation, but not SHA2-256.

If we commit to a const generic style interface, we should probably have each `Hasher<L>` implement a specific algorithm (so make it configurable by the type you pass in instead of runtime-configurable through the HIL).